Add Keywords to ProfilingSessionEventSource

pharring opened the pull request:
Enhance the ProfilingSessionEventSource with keywords and events to report Open Telemetry Resource attributes to listeners.
This lets profilers discover information from the resource attributes like the service role name.

pharring committed:
Add Keywords to ProfilingSessionEventSource
Lets listeners discover resource attributes and activity start/stop events

azure-sdk commented:
**API change check**

API changes are not detected in this pull request.

TimothyMothra commented:
We should add a test class for your EventSource class.
Sorry, I missed this in your previous PR.

Can you copy the test we have here: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Tests/AzureMonitorAspNetCoreEventSourceTests.cs
This uses a test helper that invokes all the event methods in an EventSource class and it helps us catch errors.
This is the same helper class that we used in Application Insights SDK.


Please either make a new class to match your EventSource class, or add a new method to our existing class.

pharring committed:
Spelling an nits

pharring committed:
Rename event to ResourceAttributes

pharring committed:
Fix nullability of event handler

pharring commented on sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Internals/Profiling/ProfilingSessionEventSource.cs:
```diff
@@ -37,10 +53,45 @@ internal class ProfilingSessionEventSource : EventSource
+            /// <summary>
+            /// Emits start/stop events for activities.
+            /// </summary>
+            public const EventKeywords Activities = (EventKeywords)0b_10;
```
While researching some other issues, I notice that the OpenTelemetry base SDK has its own OpenTelemetry-Sdk EventSource with ActivityStarted/Stopped.

I'm thinking, therefore, that I don't need to repeat it here. The profiler can enable that EventSource (ETW provider) instead.

pharring committed:
Drop Activities keywords and events
The profiler, instead can enable the OpenTelemetry-Sdk EventSource from the base SDK

rajkumar-rangaraj commented on sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Internals/Profiling/ProfilingSessionEventSource.cs:
```diff
@@ -83,6 +125,74 @@ protected override void OnEventCommand(EventCommandEventArgs command)
+        [NonEvent]
+        public void WriteResourceAttributes(Resource resource)
+        {
+            if (resource is null)
```
In almost all cases, a resource should have a value. Given that this is in the context of a profiler, should we simply return instead of throwing an exception?

pharring replied:
OK

rajkumar-rangaraj commented on sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Internals/Profiling/ProfilingSessionEventSource.cs:
```diff
@@ -83,6 +125,74 @@ protected override void OnEventCommand(EventCommandEventArgs command)
+            if (IsEnabled(EventLevel.Informational, Keywords.ResourceAttributes))
+            {
+                IEnumerable<KeyValuePair<string, object>> attributes = resource.Attributes;
+                ResourceAttributes(attributes.Select(kvp => new KeyValuePair<string, string>(kvp.Key, Convert.ToString(kvp.Value, InvariantCulture)!)));
```
`Convert.ToString(kvp.Value, InvariantCulture)` might throw an exception if customer overrides tostring and that throws.

pharring replied:
That is a fair point, but note that the ResourceBuilder will not let you add a non-primitive type as an attribute value ("Attribute value type is not an accepted primitive")

pharring replied:
Now that I look closer, it looks like resource attribute values can be arrays. I presume there's some serialization code for this that I can copy.

pharring replied:
I added a try/catch just in case -- although, I can't test it because the ResourceBuilder won't let me add a non-primitive type as an attribute value.

I also added length truncation.

I did not do anything for arrays. I couldn't find any existing code in the exporter that handles array attribute values. I'm thinking they're rare enough that it's OK to leave it as-is. For the profiler, the attribute values we care about (e.g. servicename) are all strings.

TimothyMothra replied:
would it make sense to only capture specific keys here?

We also have a precedent for checking the type of the value.
https://github.com/Azure/azure-sdk-for-net/blob/f65fb9760cbfd8894d40b3f02419d363690eb3bc/sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/ActivityTagsProcessor.cs#L87-L134

pharring replied:
Yeah, the attributes I'm using today are service.namespace and service.name. However, it's early days and there may be a need for others. It's easier to modify the profiler to pick out future values than it is to pre-select them here. The latter leaves us dependent on getting apps to upgrade to newer distro versions.

I didn't understand the relevance of the CategorizeTags code, @TimothyMothra. Was it to show how to string-ify arrays or a way to pick out certain kinds of attributes?

TimothyMothra replied:
> I didn't understand the relevance of the CategorizeTags code, @TimothyMothra. Was it to show how to string-ify arrays or a way to pick out certain kinds of attributes?

Both :)
But if you're comfortable with your current approach then I think it's fine to leave it as is.

pharring committed:
Safer attribute value conversion

TimothyMothra commented on sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Internals/Profiling/ProfilingSessionEventSource.cs:
```diff
@@ -83,6 +125,88 @@ protected override void OnEventCommand(EventCommandEventArgs command)
+        /// Write the attributes of a <see cref="Resource"/>.
+        /// </summary>
+        /// <param name="resource">The resource.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="resource"/> is null.</exception>
+        [NonEvent]
```
Can remove this comment. No longer throwing exception.

```suggestion
        [NonEvent]
```

TimothyMothra commented on sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Internals/Profiling/ProfilingSessionEventSource.cs:
```diff
@@ -83,6 +125,88 @@ protected override void OnEventCommand(EventCommandEventArgs command)
+            static string SafeConvertToString(object? value)
+            {
+                const int maxLength = 128;
+                try
+                {
+                    string converted = Convert.ToString(value, InvariantCulture) ?? "";
+                    return converted.Length <= maxLength ? converted : converted.Substring(0, maxLength);
+                }
+                catch
+                {
+                    return "";
+                }
```
what are you expecting to throw here?

TimothyMothra replied:
Just saw the other thread with you and Raj. Consider adding some comment here to provide context for why this is necessary.

vishweshbankwar commented on sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Internals/Profiling/ProfilingSessionEventSource.cs:
```diff
@@ -83,6 +125,88 @@ protected override void OnEventCommand(EventCommandEventArgs command)
+
+            static string SafeConvertToString(object? value)
+            {
+                const int maxLength = 128;
```
out of curiosity - why is it 128?

pharring replied:
I was trying to limit the total payload sent via EventSource. I think I've seen a maximum length of 8,192 used elsewhere in the OpenTelemetry code. That seems way too much, but perhaps I should use that.

pharring replied:
Added a comment and raised it to 200.

rajkumar-rangaraj commented on sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Internals/Profiling/ProfilingSessionEventSource.cs:
```diff
@@ -83,6 +125,88 @@ protected override void OnEventCommand(EventCommandEventArgs command)
+                }
+                catch
+                {
+                    return "";
```
nit:
```suggestion
                    return string.Empty;
```

pharring committed:
Remove exception tag from doc-comment

pharring committed:
Use string.Empty

pharring committed:
Comment SafeConvertToString

pharring committed:
Comment the max length

pharring committed:
Fix grammar

pharring commented:
If there are no further comments, This is ready to merge (I don't have permission to do it myself)
